rts/Rml/Backends/RmlUi_Backend.cpp

```diff
@@ -0,0 +1,488 @@
+{
+	if (data->debuggerAttached)
+	{
+		Rml::Debugger::Initialise(data->contexts[contextIndex]);
+		Rml::Debugger::SetVisible(true);
+	}
+	else
+	{
+		Rml::Debugger::Shutdown();
+	}
+	data->debuggerAttached = !data->debuggerAttached;
+}
+
+Rml::SystemInterface *RmlGui::GetSystemInterface()
+{
+	return &data->system_interface;
+}
+
+Rml::RenderInterface *RmlGui::GetRenderInterface()
+{
+	return &data->render_interface;
+}
+
+void createContext(const std::string &name)
+{
+	Rml::Context *context = Rml::CreateContext(name, Rml::Vector2i(data->winX, data->winY));
+	// Rml::Debugger::Initialise(context);
+	RmlGui::AddContext(context);
+}
+
+void RmlGui::CreateContext(const std::string &name)
+{
+	createContext(name);
+}
+
+void RmlGui::AddContext(Rml::Context *context)
+{
+	data->contexts.push_back(context);
+}
+
+void RmlGui::Update()
+{
+	ZoneScopedN("RmlGui Update");
+	if (!data->initialized || data->contexts.size() < 1)
+	{
+		return;
+	}
+#ifndef HEADLESS
+	for (auto &context : data->contexts)
+	{
+		context->Update();
+	}
+#endif
+}
+
+void RmlGui::RenderFrame()
+{
+	ZoneScopedN("RmlGui Draw");
+	if (!data->initialized || data->contexts.size() < 1)
+	{
+		return;
+	}
+
+#ifndef HEADLESS
+	RmlGui::BeginFrame();
+	for (auto &context : data->contexts)
+	{
+		context->Render();
+	}
+	RmlGui::PresentFrame();
+#endif
+}
+
+void RmlGui::BeginFrame()
+{
+	// data->render_interface.Clear();
+	data->render_interface.BeginFrame();
+}
+
+void RmlGui::PresentFrame()
+{
+	data->render_interface.EndFrame();
+	RMLUI_FrameMark;
+}
+
+bool processContextMouseEvent(Rml::Context *context, const SDL_Event &event)
+{
+	switch (event.type)
+	{
+	case SDL_MOUSEMOTION:
+	case SDL_MOUSEBUTTONDOWN:
+	case SDL_MOUSEBUTTONUP:
+	case SDL_MOUSEWHEEL:
+	{
+		bool handled = !RmlSDL::InputEventHandler(context, event);
+		if (!handled && event.type == SDL_MOUSEBUTTONDOWN)
+		{
+			Rml::Element *el = context->GetFocusElement();
+			if (el)
+			{
+				el->Blur();
+			}
+		}
+		data->inputReceiver.setActive(handled);
+		return handled;
+	}
+	default:
+	{
+		data->inputReceiver.setActive(false);
+		return false;
+	}
+	}
+	return false;
+}
+
+/*
+	Return true if the event was handled by rmlui
+*/
+bool RmlGui::ProcessMouseEvent(const SDL_Event &event)
+{
+	if (data == NULL || data->contexts.size() < 1 || !data->initialized)
+	{
+		return false;
+	}
+	bool result = false;
+	for (auto &context : data->contexts)
+	{
+		result |= processContextMouseEvent(context, event);
+	}
+	return result;
+}
+
+/*
+	Return true if the event was handled by rmlui
+*/
+bool RmlGui::ProcessKeyPressed(int keyCode, int scanCode, bool isRepeat)
+{
+	if (!data->initialized || data->contexts.size() < 1)
+	{
+		return false;
+	}
+	bool result = false;
+	for (auto &context : data->contexts)
+	{
+		auto kc = RmlSDL::ConvertKey(keyCode);
+		// if (kc == Rml::Input::KI_ESCAPE)
+		// {
+		// 	Rml::Element *el = context->GetFocusElement();
+		// 	if (el)
+		// 	{
+		// 		el->Blur();
```

Comment:
is this some sort of test code? remove or add instructions on when to uncomment

Reply:
I'll pull this out